src/tactic/finish.lean

```diff
@@ -409,29 +446,120 @@ open lean lean.parser interactive interactive.types
 local postfix `?`:9001 := optional
 local postfix *:9001 := many
 
-meta def clarify (hs : parse simp_arg_list) (cfg : auto_config := {}) : tactic unit :=
-do s ← mk_simp_set ff [] hs,
-   auto.clarify s cfg
+/--
+  `clarify[h1,...,hn] using [e1,...,en]` negates the goal, normalizes hypotheses
+  (by splitting conjunctions, eliminating existentials, pushing negations inwards,
+  and calling `simp` with the supplied lemmas `h1,...,hn`), and then tries `contradiction`.
+  
+  If this fails, it will create an SMT state and repeatedly use `ematch`
+  (using `ematch` lemmas in the environment, universally quantified assumptions,
+  and the supplied lemmas `e1,...,en`) and congruence closure.
 
-meta def safe (hs : parse simp_arg_list) (cfg : auto_config := {}) : tactic unit :=
-do s ← mk_simp_set ff [] hs,
-   auto.safe s cfg
+  `clarify` is complete for propositional logic.
+  
+  Either of the supplied simp lemmas or the supplied ematch lemmas are optional.
 
-meta def finish (hs : parse simp_arg_list) (cfg : auto_config := {}) : tactic unit :=
+  `clarify` will fail if it produces more than one goal.
+-/
+meta def clarify (hs : parse simp_arg_list) (ps : parse (tk "using" *> pexpr_list_or_texpr)?)
+  (cfg : auto_config := {}) : tactic unit :=
 do s ← mk_simp_set ff [] hs,
-   auto.finish s cfg
-
-meta def iclarify (hs : parse simp_arg_list) (cfg : auto_config := {}) : tactic unit :=
+   let ps' :=
+   match ps with
```

Comment:
This `match` is `ps.get_or_else []`.

Reply:
ah, didn't know about `get_or_else`---thanks for the tip!